### ISSUE TITLE
SFS and Grid Search predictions on test

### ISSUE THREAD

**Issue by ahhm14** [rasbt/mlxtend#680]
Hi all!, hope you can help me with this. 

I set up a Grid Search with sequential feature selection as it is stated in the example 8 of documentation:http://rasbt.github.io/mlxtend/user_guide/feature_selection/SequentialFeatureSelector/#example-8-sequential-feature-selection-and-gridsearch

I did a little test, with few parameters knowing that with all the parameters I want to feed the grid search grid will take a lot of training time.

It runs great, but I have a problem when making the prediction over the test set. I get a **0.91** accuracy for the train set and poor **0.36** for the test set. I believe there is something wrog because I expected a not so poor performance on the test set. 

This is my set up:

`
def SFS_GS(X_train, y_train, X_test, y_test, model, param_grid):
    
    #Setting up the SFS
    sfs1 = SFS(estimator=model, 
           k_features=X_train.shape[1],
           forward=True, 
           floating=False, 
           scoring='accuracy',
           cv=5)
    
    ### Pipeline
    
    ### we would need to adapt the "NUMBER OF FEATURES PARAMETER OF THE GRID"
    
    pipe = Pipeline([('sfs', sfs1), 
                 ('model', model)])
    
    #dict_1 = {'sfs__k_features':list(range(1,X_train.shape[1]))}   #### para pruebas
    
    dict_1 = {'sfs__k_features':[10]}  #Testing
    
    dict_1.update(param_grid)
    
    gs = GridSearchCV(estimator=pipe, 
                  param_grid=dict_1, 
                  scoring='accuracy', 
                  n_jobs=1, 
                  cv=StratifiedKFold(5),
                  verbose=3,
                  refit=True)
   
    
    gs = gs.fit(X_train, y_train)
   
    
    print("Best Model", gs.best_params_)
    
    
    y_test_pred = gs.predict(X_test)
    
    test_acc = accuracy_score(y_test,y_test_pred)
    
    print("\n Test Accuracy with best estimator: ", test_acc)`

so as you can see I use the predict of the grid search on my test dataset. 

when I run the gs best estimator I get the following:

`
Pipeline(memory=None,
         steps=[('sfs',
                 SequentialFeatureSelector(clone_estimator=True, cv=5,
                                           estimator=SVC(C=1.0, cache_size=200,
                                                         class_weight=None,
                                                         coef0=0.0,
                                                         decision_function_shape='ovr',
                                                         degree=3,
                                                         gamma='scale',
                                                         kernel='linear',
                                                         max_iter=5000,
                                                         probability=False,
                                                         random_state=42,
                                                         shrinking=True,
                                                         tol=0.001,
                                                         verbose=False),
                                           fixed_features=None, floating=False,
                                           forward=True, k_features=10,
                                           n_jobs=1, pre_dispatch='2*n_jobs',
                                           scoring='accuracy', verbose=0)),
                ('model',
                 SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
                     decision_function_shape='ovr', degree=3, gamma='scale',
                     kernel='rbf', max_iter=5000, probability=False,
                     random_state=42, shrinking=True, tol=0.001,
                     verbose=False))],
         verbose=False)`

I'm not sure which model is actually taking to make the prediction or if I missing a step before the prediction. 

Thanks a lot for your help!

PS. I have a similar set up with other feature selection techniques such as K_best and it's giving me reasonable results for the testing I performed.

**Comment by rasbt** [comment]
Hm, not sure, maybe there is some issue with how the test set was prepared. Maybe it's very different from the training set, is not normalized the same way, etc.

I would try to divide the training set into two parts. Then use 1 part for model training like you do right now. Then, use the 2nd part in place of the test set. The gap between the CV performance on part 1 and the accuracy on part 2 will give some more pointers as to what might be going on here.
